src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/CompilerGeneratedState.cs

```diff
@@ -352,27 +377,68 @@ void MapGeneratedTypeTypeParameters(MetadataType generatedType)
+                MetadataType? ScanForInit(MetadataType compilerGeneratedType, MethodIL body)
                 {
                     ILReader reader = new ILReader(body.GetILBytes());
                     while (reader.HasNext)
                     {
                         ILOpcode opcode = reader.ReadILOpcode();
+                        bool handled = false;
+                        MethodDesc? methodOperand = null;
                         switch (opcode)
                         {
                             case ILOpcode.initobj:
                             case ILOpcode.newobj:
-                                if (body.GetObject(reader.ReadILToken()) is MethodDesc { OwningType: MetadataType owningType }
-                                    && stateMachineType == owningType.GetTypeDefinition())
                                 {
-                                    return owningType;
+                                    methodOperand = body.GetObject(reader.ReadILToken()) as MethodDesc;
+                                    if (methodOperand is MethodDesc { OwningType: MetadataType owningType }
+                                        && compilerGeneratedType == owningType.GetTypeDefinition())
+                                    {
+                                        return owningType;
+                                    }
+                                    handled = true;
+                                }
+                                break;
+
+                            case ILOpcode.ldftn:
+                            case ILOpcode.ldtoken:
+                            case ILOpcode.call:
+                            case ILOpcode.callvirt:
+                                methodOperand = body.GetObject(reader.ReadILToken()) as MethodDesc;
+                                break;
+
+                            case ILOpcode.stsfld:
+                                {
+                                    if (body.GetObject(reader.ReadILToken()) is FieldDesc { OwningType: MetadataType owningType }
+                                        && compilerGeneratedType == owningType.GetTypeDefinition())
+                                    {
+                                        return owningType;
+                                    }
+                                    handled = true;
                                 }
                                 break;
 
                             default:
                                 reader.Skip(opcode);
                                 break;
                         }
+
+                        // Also look for type substitutions into generic methods
+                        // (such as AsyncTaskMethodBuilder::Start<TStateMachine>).
+                        if (!handled && methodOperand is not null)
+                        {
+                            if (methodOperand != methodOperand.GetTypicalMethodDefinition())
```

Comment:
Nit: we don't need to enter the foreach for methods on generic types, only generic methods.

```suggestion
                            if (methodOperand != methodOperand.GetMethodDefinition())
```